### DISCUSSION TITLE
PDF Text Extraction Order Not Matching Visual Layout Despite Correct Coordinates

### DISCUSSION THREAD

**Discussion by Phalgun-Santhapuri** [pymupdf/PyMuPDF#4124] [Looking for help]
I am working on extracting text from a PDF using PyMuPDF. However, I am encountering an issue where the extracted text order does not match the visual flow/Layout flow of the PDF .

Details of the Issue:

The PDF's text is correctly positioned according to its coordinates (bounding boxes), but the logical extraction order is incorrect.
For example, on the first page of my PDF:

- After extracting line 2, the tool directly jumps to a table at the bottom of the page, skipping intervening text.

- Later, it picks up lines 3–20 in an unordered manner.

- I have verified that the issue is not related to column or layout misalignment, as the coordinates are accurate.

**The document contains multi-column layouts and mixed elements like tables and have complex layouts in the PDF.**
From dict i am getting the bounding box information later i am applying the further logic.
But i observed that the the text in the dict or any other option available itself has the incorrect order of the text which i have mentioned above.

**Comment by JorjMcKie** [comment]
All of what you mention looks like normal in PDFs: in extreme cases, every single character may appear in arbitrary sequence when extracted. Only when explicitly sorting for output, a "natural" reading sequence can be established.

So we need an example page before we can say anything else.

**Reply by Phalgun-Santhapuri** [reply]
@JorjMcKie, Okay sure,
In the attached example pdf the first line is test john but when extracted the name is coming below after the table is extracted.
The text order does not match the visual layout.
**When we do extraction, we need to maintain the layout information also.**

[Pymupdf_sample.pdf](https://github.com/user-attachments/files/18072969/Pymupdf_sample.pdf)

**Reply by JorjMcKie** [reply]
This is probably for you:
```python
import pymupdf, pymupdf4llm, pathlib

doc = pymupdf.open("sample.pdf")
md = pymupdf4llm.to_markdown(doc, margins=0)
pathlib.Path(doc.name + ".md").write_text(md)
```
Produces this
[sample.pdf.md](https://github.com/user-attachments/files/18077850/sample.pdf.md)

**Reply by Phalgun-Santhapuri** [reply]
@JorjMcKie, I noticed that this solution works for the current issue. However, I observed that in cases of closely spaced text within a simple layout or column layout, the text with the smaller y-value is being printed first, which does not provide an accurate extraction. **Additionally, I noticed that text is missing from the total page content in readable documents when dealing with column-layout PDFs.**

**Reply by Phalgun-Santhapuri** [reply]
Hi @JorjMcKie ,
I wanted to follow up on my previous message regarding the issue with text extraction. I’m still awaiting your thoughts or suggestions on addressing the challenge.
Let me know if additional details are needed from my end to help with troubleshooting. Looking forward to your response!

**Reply by JorjMcKie** [reply]
There are a few situations where the latest pymupd4llm version 0.0.17 is introducing problems that were better handled in version 0.0.16.
Try to install the back level version `pip install pymupdf4llm==0.0.16`.

**Reply by Phalgun-Santhapuri** [reply]
Thank you for the response @JorjMcKie ,

I used the earlier version and noticed that in cases where there is a column layout followed by a table, the extraction method prioritizes extracting the column closest to the top border first, followed by the table content below the column layout, and then moves to the next column. However, the correct extraction process should first capture the entire column layout text before extracting the table content.

The version fails to handle complex layouts in the PDF effectively.
